streams/jvm/src/main/scala/zio/stream/experimental/platform.scala

```diff
@@ -119,4 +125,288 @@ trait ZStreamPlatformSpecificConstructors {
+                      new CompletionHandler[AsynchronousSocketChannel, Void]() {
+                        self =>
+                        override def completed(socket: AsynchronousSocketChannel, attachment: Void): Unit =
+                          callback(ZIO.succeed(Connection.make(socket)))
+
+                        override def failed(exc: Throwable, attachment: Void): Unit = callback(ZIO.fail(exc))
+                      }
+                    )
+                  }
+                  .flatMap(managedConn => registerConnection(managedConn).map(_._2))
+              }
+    } yield conn
+
+  /**
+   * Accepted connection made to a specific channel `AsynchronousServerSocketChannel`
+   */
+  class Connection(socket: AsynchronousSocketChannel) {
+
+    /**
+     * Read the entire `AsynchronousSocketChannel` by emitting a `Chunk[Byte]`
+     * The caller of this function is NOT responsible for closing the `AsynchronousSocketChannel`.
+     */
+    def read: ZStream[Any, Throwable, Byte] =
+      ZStream.unfoldChunkM(0) {
+        case -1 => ZIO.succeed(Option.empty)
+        case _ =>
+          val buff = ByteBuffer.allocate(ZStream.DefaultChunkSize)
+
+          ZIO.effectAsync[Any, Throwable, Option[(Chunk[Byte], Int)]] { callback =>
+            socket.read(
+              buff,
+              null,
+              new CompletionHandler[Integer, Void] {
+                override def completed(bytesRead: Integer, attachment: Void): Unit = {
+                  (buff: Buffer).flip()
+                  callback(ZIO.succeed(Option(Chunk.fromByteBuffer(buff) -> bytesRead.toInt)))
+                }
+
+                override def failed(error: Throwable, attachment: Void): Unit = callback(ZIO.fail(error))
+              }
+            )
+          }
+      }
+
+    /**
+     * Write the entire Chuck[Byte] to the socket channel.
+     * The caller of this function is NOT responsible for closing the `AsynchronousSocketChannel`.
+     *
+     * The sink will yield the count of bytes written.
+     */
+    def write: ZSink[Any, Throwable, Byte, Throwable, Nothing, Int] =
+      ZSink.foldLeftChunksM[Any, Throwable, Byte, Int](0) { case (nbBytesWritten, c) =>
+        ZIO.effectAsync[Any, Throwable, Int] { callback =>
+          socket.write(
+            ByteBuffer.wrap(c.toArray),
+            null,
+            new CompletionHandler[Integer, Void] {
+              override def completed(result: Integer, attachment: Void): Unit =
+                callback(ZIO.succeed(nbBytesWritten + result.toInt))
+
+              override def failed(error: Throwable, attachment: Void): Unit = callback(ZIO.fail(error))
+            }
+          )
+        }
+      }
+
+    /**
+     * Close the underlying socket
+     */
+    def close(): UIO[Unit] = ZIO.effectTotal(socket.close())
+  }
+
+  object Connection {
+
+    /**
+     * Create a `Managed` connection
+     */
+    def make(socket: AsynchronousSocketChannel): UManaged[Connection] =
```

Comment:
I would probably make the constructor and the `close` of `Connection` private to force the users to use this method